06-classification_continued.Rmd

```diff
@@ -559,16 +559,257 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+cancer_irrelevant <- read_csv("data/wdbc_irrelevant.csv") %>%
+                mutate(Class = as_factor(Class))
+
+plot_irrelev <- cancer_irrelevant %>%
+  ggplot(aes(x = Irrelevant1, y = Irrelevant2, color = Class)) +
+  geom_point(alpha = 0.5) +
+  labs(color = "Diagnosis") +
+  scale_color_manual(labels = c("Malignant", "Benign"), values = cbPalette)
+
+plot_irrelev
+```
+</center>
+
+Next, we build a sequence of K-NN classifiers that include both `Smoothness` and `Concavity` as predictor variables,
+but also increasingly many irrelevant variables. Each model is tuned via 5-fold cross-validation.
+We show the plot of estimated accuracy versus the number of irrelevant predictors below.
+<center>
+```{r 06-performance-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Effect of inclusion of irrelevant predictors"}
+set.seed(1)
+# get accuracies after including k irrelevant features
+ks <- c(0, 5, 10, 20, 50, 100, 200, 500)
+accs <- list()
+for (i in 1:length(ks)) {
+  knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
+     set_engine("kknn") %>%
+     set_mode("classification")
+
+  cancer_irrelevant_subset <- cancer_irrelevant %>% select(1:(3+ks[[i]]))
+
+  cancer_vfold <- vfold_cv(cancer_irrelevant_subset, v = 5, strata = Class)
+
+  cancer_recipe <- recipe(Class ~ ., data = cancer_irrelevant_subset) %>%
+      step_scale(all_predictors()) %>%
+      step_center(all_predictors())
+
+  
+  accs[[i]] <- workflow() %>%
+    add_recipe(cancer_recipe) %>%
+    add_model(knn_spec) %>%
+    tune_grid(resamples = cancer_vfold, grid = 10) %>%
+    collect_metrics() %>%
+    filter(.metric == "accuracy") %>%
+    summarize(mx = max(mean)) %>%
+    select(mx)
+  accs[[i]] <- accs[[i]]$mx
+}
+accs <- accs %>% unlist()
+
+# get accuracy if we always just guess the most frequent label
+base_acc <- cancer_irrelevant %>%
+                group_by(Class) %>%
+                summarize(n = n()) %>%
+                mutate(frac = n/sum(n)) %>%
+                summarize(mx = max(frac)) %>%
+                select(mx)
+base_acc <- base_acc$mx %>% unlist()
+
+# plot
+res <- tibble(ks = ks, accs = accs)
+res <- res %>% mutate(base_acc = base_acc)
+plt_irrelevant_accuracies <- res %>%
+  ggplot() +
+  geom_line(mapping = aes(x=ks, y=accs, linetype="Tuned K-NN")) +
+  geom_hline(data=res, mapping=aes(yintercept=base_acc, linetype="Always Predict Benign (63%)")) +
+  labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") + 
+  scale_linetype_manual(name="Method", values = c("dashed", "solid"))
+
+plt_irrelevant_accuracies
+```
+</center>
+
+Wow! Once we add sufficiently many irrelevant predictor variables, our classifier
+does not do much better than just always guessing "benign", 
+ the most common class label in the data (63% of observations are benign).
```

Comment:
Can you expand upon the baseline model ("Always predict Benign") here? How did you calculate this? Were irrelevant predictors added to it? I think this is the first time we have compared a model to this baseline (which I like, and now wonder if we should do it more above...)

Reply:
@ttimbers what detail are you looking for here? It just always predicts benign regardless of predictors.

You can calculate the accuracy by just computing what fraction of the training data are benign (that's the pct training accuracy you will get).

> Were irrelevant predictors added to it?

Doesn't matter if they were or not, you always predict "benign" no matter what

> which I like, and now wonder if we should do it more above...)

This is probably a really good idea. I think the right strategy would be to introduce it in the 2nd classification chapter, since that focuses on evaluation. I will leave this conversation unresolved for now, because i'll probably try to address it as part of this PR

Reply:
OK, added a discussion of this earlier on in the evaluation section